wasm/src/main/java/com/dylibso/chicory/wasm/types/ValueType.java

```diff
@@ -2,31 +2,75 @@
 
 import com.dylibso.chicory.wasm.MalformedException;
 import java.util.List;
+import java.util.Objects;
 
 /**
  * The possible WASM value types.
  */
-public enum ValueType {
-    UNKNOWN(-1),
-    F64(ID.F64),
-    F32(ID.F32),
-    I64(ID.I64),
-    I32(ID.I32),
-    V128(ID.V128),
-    FuncRef(ID.FuncRef),
-    ExternRef(ID.ExternRef);
+public final class ValueType {
+    private static final int EMPTY_OPERAND = 0;
+    private static final long OPCODE_MASK = 0x0000FFFFL;
+    private static final long OPERAND_SHIFT = 32;
 
-    private final int id;
+    public static ValueType UNKNOWN = new ValueType(ValueTypeOpCode.UNKNOWN);
+    public static ValueType F64 = new ValueType(ValueTypeOpCode.F64);
 
-    ValueType(int id) {
-        this.id = id;
+    public static ValueType F32 = new ValueType(ValueTypeOpCode.F32);
+    public static ValueType I64 = new ValueType(ValueTypeOpCode.I64);
+
+    public static ValueType I32 = new ValueType(ValueTypeOpCode.I32);
+
+    public static ValueType V128 = new ValueType(ValueTypeOpCode.V128);
+    public static ValueType FuncRef = new ValueType(ValueTypeOpCode.FuncRef);
+    public static ValueType ExternRef = new ValueType(ValueTypeOpCode.ExternRef);
+
+    private final Variant opcode;
+
+    // some value types have an argument, conveniently this fits in an int
+    private final int operand;
+
+    public ValueType(ValueTypeOpCode opcode) {
+        this(opcode, EMPTY_OPERAND);
+    }
+
+    public ValueType(ValueTypeOpCode opcode, int operand) {
+        Variant var = Variant.ofValueTypeOpCode(opcode);
+
+        // handle operand
+        switch (opcode) {
+            case FuncRef:
+                assert operand == EMPTY_OPERAND;
+                operand = OperandCode.FUNC.code();
+                break;
+            case ExternRef:
+                assert operand == EMPTY_OPERAND;
+                operand = OperandCode.EXTERN.code();
+                break;
+            case RefNull:
+            case Ref:
+                assert operand == OperandCode.FUNC.code()
+                        || operand == OperandCode.EXTERN.code()
+                        || operand >= 0;
+                break;
+            default:
+                assert operand == EMPTY_OPERAND;
+        }
+        this.opcode = var;
+        this.operand = operand;
     }
 
     /**
-     * @return the numerical identifier for this type
+     * @return the numerical identifier for this type. Conveniently, all value types we want to represent
+     *     for now (constructor + arguments) can fit inside a Java long.
+     *
+     *     We store as operand in the MSB and the opcode in the LSB.
```

Comment:
This is fine for me, but I'm a little confused, either we keep the "operand" in the `id` or as a separate field, but not both.
Any specific reason for this choice?